ДЗ 12

DaniilKosmachev opened the pull request.

DaniilKosmachev committed:
ДЗ 12, вариант 1:
1) История поиска (сохранение в SharedPreferences и чтение из него реализовано в отдельном классе, который принимает объект SharedPreferences)
2) Использован один и тоже ViewHolder, но с разными адаптерами
3) В Адаптере на itemView установлен слушатель
4) В SearchActivity.kt оба адаптера со слушателями (один боевой - добавляет в историю поиска, другой Toast, для контроля работы слушателя)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/App.kt:
```diff
@@ -0,0 +1,33 @@
+    private lateinit var themeSwitchSharedPreferences: SharedPreferences
+    private lateinit var classThemeSwitch: AppThemeSwitch
+
+    companion object {
```
Согласно конвенции котлина, companion object [должен находиться в конце класса](https://kotlinlang.org/docs/coding-conventions.html#class-layout)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -74,11 +85,60 @@ companion object {
         onUpdateButtonClickListener()
-        trackAdapter = TrackAdapter(iTunesTrack)
-        recycleViewTrack.adapter = trackAdapter
+        initizlizedComponents()
```
здесь опечатка, корректнее initializeComponents

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -74,11 +85,60 @@ companion object {
+    }
+
+
+    private fun addNewTrackInTrackHistory(newTrack: Track) {
```
Эту логику можно было бы вынести из активити во вспомогательный класс, чтобы не было перегружено

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -74,11 +85,60 @@ companion object {
+        val temporaryTrackArray = ArrayList<Track>()
+        temporaryTrackArray.addAll(classHistorySearch.getTrackArrayFromShared())//получаем десериализованный список треков из SP во временный лист
+        if (temporaryTrackArray.isEmpty()) {//если он пуст, то просто добавляем новый трек на нулевой индекс
+            iTunesTrackSearchHistory.add(0,newTrack)
```
Можно не писать индекс, если лист пустой, элемент и так добавится первым (он же будет и последним в этой ситуации)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -74,11 +85,60 @@ companion object {
+        if (temporaryTrackArray.isEmpty()) {//если он пуст, то просто добавляем новый трек на нулевой индекс
+            iTunesTrackSearchHistory.add(0,newTrack)
+        } else if (temporaryTrackArray.isNotEmpty()) {
+            if (temporaryTrackArray.size == 11) {
```
11 и 10 в данном случае [магическое число](https://ru.wikipedia.org/wiki/%D0%9C%D0%B0%D0%B3%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%BE%D0%B5_%D1%87%D0%B8%D1%81%D0%BB%D0%BE_(%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)).  Всегда лучше выносить такие в константы)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -74,11 +85,60 @@ companion object {
+                if (currentTrack.trackId == newTrack.trackId) {
+                    iterator.remove()
+                }
+            }
```
Можно заменить одной строкой  
```suggestion
            temporaryTrackArray.removeIf { it.trackId == newTrack.trackId }
```

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -111,9 +175,36 @@ companion object {
+    private fun setOnEditTextFocusLisneter() {
+        editTextSearch.setOnFocusChangeListener { _, hasFocus ->
+            if (iTunesTrackSearchHistory.isNotEmpty()) {
+            historyLayout.visibility = if (hasFocus && editTextSearch.text.isNullOrEmpty()) View.VISIBLE else View.GONE
```
Здесь не хватает отступов, чтобы код визуально был внутри if. Поможет комбинация автоматического форматирования ctrl + alt + L или cmd + opt + L

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/SettingsActivity.kt:
```diff
@@ -1,36 +1,64 @@
         writeToSupportLinearClickListener()
         shareAppLinearClickListener()
         userContractLinearClickListener()
+        themeSwitchClickListener()
+        initializedVariable()
```
Названия методов лучше писать по формуле: глагол в инфинитивной форме + существительные. [Рекомендации от конвенции](https://kotlinlang.org/docs/coding-conventions.html#choose-good-names:~:text=The%20name%20of%20a%20method,sorted%20copy%20of%20the%20collection.)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/SettingsActivity.kt:
```diff
@@ -1,36 +1,64 @@
-    private fun backToMainActivityClickLisnter() {
+    private fun themeSwitchClickListener() {
+        themeSwitch.setOnCheckedChangeListener { _, checked ->
+            (applicationContext as App).switchTheme(checked)
```
Можно было бы сократить код в этом классе, если бы запись в sharedPreferences осуществлялась внутри метода App.switchTheme()

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/Track.kt:
```diff
@@ -3,5 +3,6 @@ package com.example.playlistmaker
 data class Track (val trackName: String,
```
Здесь тоже форматирование немного поехало. Если полей много, то лучше их писать с новой строки, начиная с первого

Anna-Malinina commented on app/src/main/res/layout/activity_library.xml:
```diff
@@ -8,13 +8,14 @@
-            android:text="Это ативити Медиатеки!"
-            android:textSize="25dp"
+            android:gravity="center"
+            android:text="Это активити Медиатеки!"
```
Текст можно вынести в ресурсы. Даже не смотря на то, что это временная надпись. Это помогает набивать руку на хорошие практики

Anna-Malinina commented on app/src/main/res/layout/track_view.xml:
```diff
@@ -5,7 +5,9 @@
     xmlns:app="http://schemas.android.com/apk/res-auto"
     android:orientation="horizontal"
-    android:paddingLeft="@dimen/main_padding_left_image">
+    android:paddingLeft="@dimen/main_padding_left_image"
```
Google не рекомендует использовать аттрибуты Left/Right, потому что экран будет некорректно отображаться при языках, которые читаются справа налево (иврит, арабский языки и т.д). [Ссылка](https://developer.android.com/about/versions/android-4.2.html#RTL). Вместо этого лучше использовать start/end.

DaniilKosmachev committed:
ДЗ 12, вариант 2:
1) Учел почти все рекоммендации ревьера
2) Чутка поправил верстку - вместо ImageButton поставил ImageView в track_view.xml и activity_settings.xml
